src/cloudformation_cli_python_lib/resource.py

```diff
@@ -73,8 +121,12 @@ def _invoke_handler(
             return ProgressEvent.failed(
                 HandlerErrorCode.InternalFailure, f"No handler for {action}"
             )
-
-        return handler(session, request, callback_context)
+        progress = handler(session, request, callback_context)
+        is_in_progress = progress.status == OperationStatus.IN_PROGRESS
+        is_mutable = action in MUTATING_ACTIONS
+        if is_in_progress and not is_mutable:
+            raise Exception("READ and LIST handlers must return synchronously.")
```

Comment:
could return a progress event with internal failure, as that will effectively be the same? or a custom exception?